entropy/windows/Raaz/Entropy.hsc

```diff
@@ -38,28 +35,22 @@ foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptAcquireContextW"
 foreign import WINDOWS_CCONV unsafe "Wincrypt.h CryptReleaseContext"
     c_CryptReleaseContext :: HCRYPTPROV -> Word32 -> IO Bool
 
--- | Cache the crypto context so we don't have to create it on each call.
-cryptoContext :: ForeignPtr HCRYPTPROV
-cryptoContext = unsafePerformIO $
-  do buffer <- mallocBytes (#size HCRYPTPROV)
-     addr   <- new buffer
-     ctx    <- newForeignPtr addr (freeContext buffer >> free addr)
-     ctx_ok <- withForeignPtr ctx $ \ptr ->
-                    c_CryptAcquireContext ptr nullPtr nullPtr
-                        (#const PROV_RSA_FULL)
-                        ((#const CRYPT_VERIFYCONTEXT) .|. (#const CRYPT_SILENT))
-     if ctx_ok
-        then return ctx
-        else error "Call to CryptAcquireContext failed."
-
--- | Release the crytographical context handle.
-freeContext :: HCRYPTPROV -> IO ()
-freeContext ctx = c_CryptReleaseContext ctx 0 >> return ()
-
 -- | Get cryptographically random bytes from the system.
 getEntropy :: (MonadIO m, LengthUnit l) => l -> Pointer -> m (BYTES Int)
-getEntropy l ptr = liftIO $ withForeignPtr cryptoContext $ \ctx ->
-    do ctx' <- peek ctx
+getEntropy l ptr = liftIO $
+    -- You can't get an address of a pointer in haskell, so making out
+    -- parameters are messy. We do this by boxing the type.
+    do ctx    <- mallocBytes (#size HCRYPTPROV)
```

Comment:
I guess the code outline should be like this.

```haskell
getEntropy = allocaBuffer (sizeOf (undefined :: Ptr HCRYPTPROV)) \ ptr -> 
     do _ <- peek ptr nullPtr
         cryptoAcquireCxt (castPtr ptr) ....
         cxt <- peek (castPtr ptr)
         do the reading
          cryptoReleaseCxt cxt
         



```

Just for reference here is what the entropy package does.

https://github.com/TomMD/entropy/blob/master/System/EntropyWindows.hs